src/Parser/Rogue/Subtlety/Modules/ResourceTracker/ResourceTracker.js

```diff
@@ -0,0 +1,105 @@
+import Analyzer from 'Parser/Core/Analyzer';
+import Combatants from 'Parser/Core/Modules/Combatants';
+
+class ResourceTracker extends Analyzer {
+  static dependencies = {
+    combatants: Combatants,
+  };  
+
+  generated = 0;
+  wasted = 0;
+  spent = 0;
+  spendersCasts = 0;
+  current = 0;
+  resourceType;
+  resourceName;
+
+  // stores resource gained/spent/wasted per ability ID
+  buildersObj = {};
+  spendersObj = {};
+  
+  initBuilderAbility(spellId) {        
+    this.buildersObj[spellId] = { generated: 0, wasted: 0 };
+  }
+  
+  initSpenderAbility(spellId) {        
+    this.spendersObj[spellId] = { spent: 0, casts: 0 };
+  }
+  
+  //Builders
+  on_toPlayer_energize(event) {
+    const spellId = event.ability.guid;
+
+    if(!this.shouldProcessEnergizeEvent(event)) {
+        return;
+    }
+
+    if (!(spellId in this.buildersObj)) {
+        this.initBuilderAbility(spellId);
+    }
+
+    this.applyBuilder(event);
+  }
+  
+  applyBuilder(event) {
+    const spellId = event.ability.guid;
+    const waste = event.waste;
+    const gain = event.resourceChange - waste;
+    
+    this.buildersObj[spellId].wasted += waste;
+    this.buildersObj[spellId].generated += gain;
+    this.wasted += waste;
+    this.generated += gain;  
+
+    const eventResource = this.getResource(event);
+    if(eventResource !== undefined) {
+      this.current = eventResource.amount;
+    }
+    else {
+      this.current += gain;
+    }
+  }
+
+  shouldProcessEnergizeEvent(event) {
+    return event.resourceChangeType === this.resourceType;
+  }  
+
+  //Spenders
+  on_byPlayer_cast(event) {
+    const spellId = event.ability.guid;
+
+    if(!this.shouldProcessCastEvent(event)) {
+        return;
+    }
+    
+    if (!(spellId in this.spendersObj)) {
+      this.initSpenderAbility(spellId);
+  }
+
+    const eventResource = this.getResource(event);
+
+    const cost = eventResource.cost;
```

Comment:
this will crash if the class resource isn't found in the event since it only checks for `undefined` while `.find` returns `null` upon no result.

Reply:
defiantly does not crash, it is used for both CPs and Energy, and CP resource is absent on most cast events

Reply:
It will crash if there's ever an instance where the energy resource is deactivated.